lib/assert.js

```diff
@@ -481,18 +479,32 @@ function partiallyCompareSets(actual, expected, comparedObjects) {
 
   if (isDeepEqual === undefined) lazyLoadComparison();
 
-  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
+  const actualSet = new SafeSet(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
   const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
-  const usedIndices = new SafeSet();
 
-  expectedIteration: for (const expectedItem of expectedIterator) {
-    for (let actualIdx = 0; actualIdx < actualArray.length; actualIdx++) {
-      if (!usedIndices.has(actualIdx) && isDeepStrictEqual(actualArray[actualIdx], expectedItem)) {
-        usedIndices.add(actualIdx);
-        continue expectedIteration;
+  for (const expectedItem of expectedIterator) {
+    let foundMatch = false;
+
+    // Check for primitives first to avoid useless loops
+    if (isPrimitive(expectedItem)) {
+      if (actualSet.has(expectedItem)) {
+        actualSet.delete(expectedItem);
```

Comment:
This could be improved further by just adding entries that are not detected. So we just lazily allocate the set for objects that are not reference identical on the other side.

This is how it's implemented in the current deepStrictEqual comparison.